### DISCUSSION TITLE
Proof verification fails for ONNX models

### DISCUSSION THREAD

**Discussion by guyu96** [zkonduit/ezkl#296] [Q&A] [answered]
Steps to repro:

1. Do a fresh build of ezkl
2. `cd examples/onnx/self_attention`
3. Generate SRS with: `ezkl gen-srs --logrows 17 --params-path=kzg.params`
4. Setup the circuit with `ezkl setup -M network.onnx --params-path=kzg.params --vk-path=vk.key --pk-path=pk.key --circuit-params-path=circuit.params`
5. With the pk/vk/circuit_params generated, prove with `ezkl prove -M network.onnx -D input.json --pk-path=pk.key --proof-path=model.proof --params-path=kzg.params --circuit-params-path=circuit.params`

Then the proof fails with constraints unsatisfied. The stdout looks something like this:

```
[*] [0s, ezkl] -
 |
 |          ███████╗███████╗██╗  ██╗██╗
 |          ██╔════╝╚══███╔╝██║ ██╔╝██║
 |          █████╗    ███╔╝ █████╔╝ ██║
 |          ██╔══╝   ███╔╝  ██╔═██╗ ██║
 |          ███████╗███████╗██║  ██╗███████╗
 |          ╚══════╝╚══════╝╚═╝  ╚═╝╚══════╝
 |
 |          -----------------------------------------------------------
 |          Easy Zero Knowledge Learning.
 |          -----------------------------------------------------------
 |
 |
[*] [0s, ezkl] - command:
 |   {
 |    "command": {
 |      "Prove": {
 |        "check_mode": "SAFE",
 |        "circuit_params_path": "circuit.params",
 |        "data": "input.json",
 |        "model": "network.onnx",
 |        "params_path": "kzg.params",
 |        "pk_path": "pk.key",
 |        "proof_path": "model.proof",
 |        "strategy": "Single",
 |        "transcript": "Blake"
 |      }
 |    }
 |  }
[*] [0s, ezkl_lib::graph::model] - set batch size to 1
[*] [0s, ezkl_lib::graph::model] - calculating num of constraints using dummy model layout...
[*] [8s, ezkl_lib::graph::model] - The model generates 4129702 constraints (does not include modules)
[*] [8s, ezkl_lib::graph] - public inputs lengths: [4096]
[*] [8s, ezkl_lib::pfsys] - loading params from "kzg.params"
[*] [8s, ezkl_lib::execute] - downsizing params to 17 logrows
[*] [8s, ezkl_lib::pfsys] - loading proving key from "pk.key"
[*] [8s, ezkl_lib::graph::model] - configuring model
[*] [32s, ezkl_lib::graph::model] - configuring model
[*] [32s, ezkl_lib::graph::model] - model layout...
[*] [41s, ezkl_lib::graph::model] - computing...
[*] [98s, ezkl_lib::graph::model] - computing...
[E] [162s, ezkl] - failed: verification failed
```

I suspect step 4 is where things went wrong, since in ezkl's docs the switch `-D input.json` is also required. However, in the latest ezkl build, `-D` is no longer a valid switch.

**Comment by jasonmorton** [comment, marked as answer]
We don't use `-D` in setup anymore, and the docs were [recently updated](https://docs.ezkl.xyz/command_line_interface/), please let us know if this is still out there somewhere. I think the issue here is that you will need to adjust the scale (to 5), bits (to 23), and logrows (to 24) from the defaults for this model. We should make that clear in the example! Thanks for pointing it out.

To do this, download (or generate) your srs with k=24
```bash
wget  https://trusted-setup-halo2kzg.s3.eu-central-1.amazonaws.com/perpetual-powers-of-tau-raw-24
```
then setup
```bash
ezkl setup -S 5 -B 23 -K 24 -M network.onnx --params-path perpetual-powers-of-tau-raw-24 --vk-path sa.vk --pk-path sa.pk --circuit-params-path sa.cpp
```
and prove.
